pkg/logging/logger.go

```diff
@@ -0,0 +1,237 @@
+	return logger
+}
+
+// NewLogger returns a new *Logger with the provided log level.
+func NewLogger(level Level) (*Logger, error) {
+	return newLoggerInternal(level, nil)
+}
+
+// NewLoggerOrDie returns a new *Logger with the provided log level or panics if there is an error
+// configuring the logger
+func NewLoggerOrDie(level Level) *Logger {
+	logger, err := newLoggerInternal(level, nil)
+	if err != nil {
+		panic(err)
+	}
+	return logger
+}
+
+func newLoggerInternal(level Level, writer zapcore.WriteSyncer) (*Logger, error) {
+	cfg := zap.NewProductionConfig()
+	if level >= DebugLevel {
+		cfg = zap.NewDevelopmentConfig()
+	}
+	cfg.EncoderConfig.EncodeTime = func(time time.Time, encoder zapcore.PrimitiveArrayEncoder) {
+		zapcore.RFC3339TimeEncoder(time.UTC(), encoder)
+	}
+	cfg.EncoderConfig.EncodeLevel = zapcore.CapitalLevelEncoder
+	cfg.DisableStacktrace = false
+	cfg.Level = zap.NewAtomicLevelAt(toZapLevel(level))
+
+	options := []zap.Option{}
+	if writer != nil {
+		var encoder zapcore.Encoder
+		// As far as I can tell from looking at the source code, there aren't constants for this and
+		// we can't pull the current encoder from the `core` passed to `WrapCore` So we have to do
+		// this a bit hackily by checking the config
+		switch cfg.Encoding {
+		case "json":
+			encoder = zapcore.NewJSONEncoder(cfg.EncoderConfig)
+		case "console":
+			encoder = zapcore.NewConsoleEncoder(cfg.EncoderConfig)
+		default:
+			return nil, fmt.Errorf("unknown encoding: %s", cfg.Encoding)
+		}
+		options = append(options, zap.WrapCore(func(core zapcore.Core) zapcore.Core {
+			// We create a new core so that we can use the provided writer. `core` is technically
+			// wrapped to take advantage of its `LevelEnabler` functionality
+			return zapcore.NewCore(encoder, writer, core)
+		}))
+	}
+	zl, err := cfg.Build(options...)
+	if err != nil {
+		return nil, err
+	}
+
+	return Wrap(zapr.NewLoggerWithOptions(zl)), nil
+}
+
+// ContextWithLogger returns a context.Context that has been augmented with
+// the provided *Logger.
+func ContextWithLogger(ctx context.Context, logger *Logger) context.Context {
+	return context.WithValue(ctx, loggerContextKey{}, logger)
+}
+
+// LoggerFromContext extracts a *Logger from the provided context.Context and
+// returns it. If no *Logger is found, a global *Logger is returned.
+func LoggerFromContext(ctx context.Context) *Logger {
+	if logger := ctx.Value(loggerContextKey{}); logger != nil {
+		return ctx.Value(loggerContextKey{}).(*Logger) // nolint: forcetypeassert
+	}
+	return globalLogger
+}
+
+// Error logs a message at the error level.
+func (l *Logger) Error(err error, msg string, keysAndValues ...any) {
+	l.callStackHelper()
+	l.logger.Error(err, msg, keysAndValues...)
+}
+
+// Warn logs a message at the warn level.
+func (l *Logger) Warn(msg string, keysAndValues ...any) {
+	// NOTE(thomastaylor312): Yep, logr be dumb. There is no WARN, so it just shows up as Info. If
```

Comment:
If there were one log level I wouldn't care about losing entirely, it would be warn. In general, people don't pay adequate heed to warnings. If it's important enough to someone _should_ pay attention, it maybe ought to be an error. If it's innocent enough to safely ignore, then it can be info.

I would be equally fine with keeping it or losing it. Dealer's choice.

Either way, this whole PR LGTM.

I will ping you offline about one possible, optional feature that could be added with about three lines of code.